hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PromotionCommitOperation.java

```diff
@@ -176,17 +196,36 @@ private void finalizePromotion() {
                     + ". Promotion partition state version: " + partitionState.getVersion()
                     + ", current partition state version: " + partitionService.getPartitionStateVersion());
         }
+
+        PromotionOperationCallback finalizePromotionsCallback = new FinalizePromotionOperationCallback(this, promotions.size());
+
         for (MigrationInfo promotion : promotions) {
             if (logger.isFinestEnabled()) {
                 logger.finest("Submitting FinalizePromotionOperation for promotion: " + promotion + ". Result: " + success);
             }
-            FinalizePromotionOperation op = new FinalizePromotionOperation(promotion, success);
+            Operation op = new FinalizePromotionOperation(promotion, success, finalizePromotionsCallback);
             op.setPartitionId(promotion.getPartitionId()).setNodeEngine(nodeEngine).setService(partitionService);
             operationService.execute(op);
         }
-        partitionService.getMigrationInterceptor()
-                .onPromotionComplete(MigrationParticipant.DESTINATION, promotions, success);
-        partitionService.getMigrationManager().releasePromotionPermit();
+    }
+
+    private void complete() {
+        InternalPartitionServiceImpl service = getService();
+        service.getMigrationInterceptor().onPromotionComplete(MigrationParticipant.DESTINATION, promotions, success);
+        PartitionEventManager eventManager = service.getPartitionEventManager();
+        MigrationStateImpl ms = migrationState;
+        for (MigrationInfo promotion : promotions) {
+            ms = ms.onComplete(1, 0L);
```

Comment:
Minor: why don't we define the `elapsed` here?

Reply:
Promotions are executed mostly concurrently, on different partitions. So sum of individual elapsed times doesn't match the total elapsed time. Also promotions are local, I think it won't be a significant time. That's why I just ignored it.

Reply:
Got it, thanks!